File: MATLAB/functions/FINDUtilde.m
function [params, y, U_bar, scalist, varlist, iter, density] = FINDUtilde(params, fund, conv_params, mu, initial)
%FINDUTILDE Recover the rural-utility primitive for a target urban share.
if nargin < 5, initial = []; end % Use original equilibrium guesses if unspecified.
progress_clock = tic; % Time progress updates independently of the numerical iteration.
for iter = 1:conv_params.max_iter % Bound the number of outer inversion steps.
    [y,U_bar,scalist,varlist,~,density] = EQFIND(params,fund,conv_params,initial); % Solve for the current rural primitive.
    initial = struct('y',y,'U_bar',U_bar,'density',density); % Explicitly carry forward the solved state.
    rel_error = abs((scalist.N/params.N_bar-mu)/max(abs(mu),1e-8)); % Relative urban-share mismatch.
    if toc(progress_clock) >= 5
        STATUS(conv_params,'Matching urbanization: iteration %d, current %.2f%%, target %.2f%%.',iter,100*scalist.N/params.N_bar,100*mu); % Reassure users during long moment matching.
        progress_clock = tic; % Limit output to approximately one update every five seconds.
    end
    if rel_error < conv_params.tol_U
        return; % Return the actual primitive used in this equilibrium, without a final unmatched update.
    end
    U_tilde_up = U_bar.^params.zeta .* ((1-mu)/mu); % Invert the rural/urban population-choice equation.
    conv_U = conv_params.conv_U ./ (1+exp(-1000.*abs(U_tilde_up-params.U_tilde)+1)); % Replication damping.
    if conv_params.type == "rich"
        conv_U = conv_params.A_U+(conv_U-conv_params.A_U)/(1+exp(-conv_params.B_U*(iter-conv_params.M_U)))^(1/conv_params.v_U); % Replication early acceleration.
    elseif conv_params.type == "jump" && iter <= conv_params.jump_n_U
        conv_boost = min(1,(conv_params.jump_m_U*iter+conv_params.jump_x_U*conv_params.max_iter)./((1+conv_params.jump_x_U)*conv_params.max_iter)); % Alternative acceleration exponent.
        conv_U = conv_U.^conv_boost; % Apply the chosen early weight.
    end
    params.U_tilde = conv_U*U_tilde_up+(1-conv_U)*params.U_tilde; % Update only the rural primitive.
end
error('ABSJ2027:Inversion','Rural-utility inversion did not match urbanization after %d iterations (relative error %.3g).',iter,rel_error); % Surface failure.
end

File: Python/functions/find_utilde.py
"""Urbanization inversion matching MATLAB FINDUtilde.m."""
import numpy as np
from eqfind import eqfind


def find_utilde(params, fund, options, urban_share, initial=None):
    params = dict(params)
    state = initial
    for iteration in range(1, options["max_iter"] + 1):
        y, utility, scalist, varlist, _, density = eqfind(params, fund, options, state)
        state = {"y": y, "U_bar": utility, "density": density}
        error = abs((scalist["N"] / params["N_bar"] - urban_share) / urban_share)
        if error < options["tol_U"]:
            return params, y, utility, scalist, varlist, iteration, density
        target = utility ** params["zeta"] * (1 - urban_share) / urban_share
        weight = options["conv_U"] / (1 + np.exp(-1000 * abs(target - params["U_tilde"]) + 1))
        if options["type"] == "rich":
            denominator = (1 + np.exp(-options["B_U"] * (iteration - options["M_U"]))) ** (1 / options["v_U"])
            weight = options["A_U"] + (weight - options["A_U"]) / denominator
        params["U_tilde"] = weight * target + (1 - weight) * params["U_tilde"]
    raise RuntimeError("Rural-utility inversion did not converge.")
